typing/includemod_errorprinter.ml

```diff
@@ -915,18 +911,41 @@ let report_apply_error ~loc env (lid_app, mty_f, args) =
       in
       Location.errorf ~loc "%t" (Functor_suberror.App.single_diff g e more)
   | _ ->
-      let actual = Functor_suberror.App.got d in
-      let expected = Functor_suberror.expected d in
-      let sub =
-        List.rev @@
-        Functor_suberror.params functor_app_diff env ~expansion_token:true d
+      let not_functor =
+        List.for_all (function _, Diffing.Delete _ -> true | _ -> false) d
       in
-      Location.errorf ~loc ~sub
-        "@[<hv>The functor application %tis ill-typed.@ \
-         These arguments:@;<1 2>\
-         @[%t@]@ do not match these parameters:@;<1 2>@[functor@ %t@ -> ...@]@]"
-        may_print_app
-        actual expected
+      if not_functor then
+        match lid_app with
+        | Some lid ->
+            Location.errorf ~loc
+              "@[The module %a is not a functor, it cannot be applied.@]"
+              Printtyp.longident lid
+        |  None ->
+            Location.errorf ~loc
+              "@[This module is not a functor, it cannot be applied.@]"
+      else
+        let intro ppf =
+          match lid_app with
+          | None -> Format.fprintf ppf "This functor application is ill-typed."
+          | Some (Longident.Lapply _ as lid) ->
+              Format.fprintf ppf "The functor application %a is ill-typed."
+                Printtyp.longident lid
```

Comment:
I don't understand the code here. The error message suggests that what is shown is the full functor application (so of the form `Foo(Bar)`), but the code suggests that `lid`in fact only denotes the functor part of the application (so only `Foo`), in the case where it is an application itself. So I would expect the error message to be "off", with a problematic application `((F(x))(Y)` being reported as an error on `F(X)`. But in fact the testsuite seems to suggest that the implementation is correct. What am I missing about what `lid_app` means?

Reply:
`lid_app` is a "best"-effort nominal approximation for the functor application. In particular it is either:

* `Some (Lapply _ )` if the functor application can be resumed as a path-with-application `F(X(A).B)(Y)`, in practice we only recover those in the easy case of functor application at the type level (`F(X).t`) 
* `Some (<Module path of the leftmost functor>)` if we are not in the case above but at least the leftmost functor has a name
* `None` if there are no name for neither the functor application nor the leftmost functor

I propose to split this option into a specialized ternary variant rather than rely on its content to distinguish the two approximation cases.

Reply:
Yes please!